cpp/src/detail/permute_range.cu

```diff
@@ -0,0 +1,191 @@
+                                           d_target_ranks.size(),
+                                           vertex_t{0},
+                                           vertex_t{comm_size},
+                                           rng_state);
+
+      thrust::sort_by_key(handle.get_thrust_policy(),
+                          d_target_ranks.begin(),
+                          d_target_ranks.end(),
+                          permuted_intergers.begin());
+
+      rmm::device_uvector<vertex_t> d_reduced_ranks(comm_size, handle.get_stream());
+      rmm::device_uvector<vertex_t> d_reduced_counts(comm_size, handle.get_stream());
+
+      auto output_end = thrust::reduce_by_key(handle.get_thrust_policy(),
+                                              d_target_ranks.begin(),
+                                              d_target_ranks.end(),
+                                              thrust::make_constant_iterator(1),
+                                              d_reduced_ranks.begin(),
+                                              d_reduced_counts.begin(),
+                                              thrust::equal_to<int>());
+
+      auto nr_output_pairs =
+        static_cast<vertex_t>(thrust::distance(d_reduced_ranks.begin(), output_end.first));
+
+      std::vector<vertex_t> h_reduced_ranks(comm_size);
+      std::vector<vertex_t> h_reduced_counts(comm_size);
+
+      raft::update_host(
+        h_reduced_ranks.data(), d_reduced_ranks.data(), nr_output_pairs, handle.get_stream());
+
+      raft::update_host(
+        h_reduced_counts.data(), d_reduced_counts.data(), nr_output_pairs, handle.get_stream());
+
+      for (int i = 0; i < static_cast<int>(nr_output_pairs); i++) {
+        tx_value_counts[h_reduced_ranks[i]] = static_cast<size_t>(h_reduced_counts[i]);
+      }
+    }
+
+    std::tie(permuted_intergers, std::ignore) = cugraph::shuffle_values(
+      handle.get_comms(), permuted_intergers.begin(), tx_value_counts, handle.get_stream());
+
+    // shuffle/permute locally again
+    fractional_random_numbers.resize(permuted_intergers.size(), handle.get_stream());
+
+    cugraph::detail::uniform_random_fill(handle.get_stream(),
+                                         fractional_random_numbers.data(),
+                                         fractional_random_numbers.size(),
+                                         float{0.0},
+                                         float{1.0},
+                                         rng_state);
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        fractional_random_numbers.begin(),
+                        fractional_random_numbers.end(),
+                        permuted_intergers.begin());
+
+    // take care of deficits and extras numbers
+
+    int nr_extras =
+      static_cast<int>(permuted_intergers.size()) - static_cast<int>(local_range_size);
+    int nr_deficits = nr_extras >= 0 ? 0 : -nr_extras;
```

Comment:
`int` can possibly overflow for very large graphs in the future (local range size can possibly exceed 2^31-1).

```
size_t nr_extras{0};
size_t nr_deficits{0};
if (permuted_integers.size() > static_cast<size_t>(local_range_size) {
  nr_extras = permuted_integers.size() - static_cast<size_t>(local_range_size);
}
else {
  nr_deficits = static_cast<size_t>(local_range_size) - permuted_integers.size();
}
```